### ISSUE TITLE
IotHubException "Work must be complete in the same order as it was started. Expected work id: '2', actual work id: '9522' "

### ISSUE THREAD

**Issue by wickste** [Azure/azure-iot-sdk-csharp#853]
- **OS, version, SKU and CPU architecture used:** Windows 10 Enterprise 1809 x64
- **Application's .NET Target Framework :** Targeting UWP 17763
- **Device:** Laptop
- **SDK version used:** Microsoft.Azure.Devices.Client 1.19.0 (also repros with 1.29.0-preview-002)

## Description of the issue:

When using ModuleClient in a UWP device client application I am getting the following exception when sending a message with several properties. The app is intended to run on Windows 10 IoT Core device, and the bug also repros when run on PC (i.e. no need to set up device to repro). The same message can be sent successfully using DeviceClient. Also it works from a plain console app with both Module and Device client.
```
An exception of type 'Microsoft.Azure.Devices.Client.Exceptions.IotHubException' occurred in System.Private.CoreLib.dll but was not handled in user code
Work must be complete in the same order as it was started. Expected work id: '2', actual work id: '9522'

   at Microsoft.Azure.Devices.Client.Transport.Mqtt.OrderedTwoPhaseWorkQueue`2.CompleteWorkAsync(IChannelHandlerContext context, TWorkId workId)
   at Microsoft.Azure.Devices.Client.Transport.Mqtt.MqttIotHubAdapter.ProcessMessage(IChannelHandlerContext context, Packet packet)
   at Microsoft.Azure.Devices.Client.Transport.Mqtt.MqttIotHubAdapter.SendMessageAsync(IChannelHandlerContext context, Message message)
   at Microsoft.Azure.Devices.Client.Transport.Mqtt.MqttIotHubAdapter.WriteAsync(IChannelHandlerContext context, Object data)
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.<>c__DisplayClass27_0.<<ExecuteWithErrorHandlingAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync[T](Func`1 asyncOperation)
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.<>c__DisplayClass18_0.<<SendEventAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.SendEventAsync(Message message, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.InternalClient.SendEventAsync(Message message)
   at ModuleClientTest.MainPage.<>c__DisplayClass3_0.<<MainPage_Loaded>b__0>d.MoveNext()
```

## Code sample exhibiting the issue:

```C#
private async void MainPage_Loaded(object sender, RoutedEventArgs e)
{
    ModuleClient client1 = ModuleClient.CreateFromConnectionString("<insert_module_connection_string>", TransportType.Mqtt);
    client1.SetConnectionStatusChangesHandler(ConnectionStatusChanged);
    await client1.OpenAsync().ConfigureAwait(true);
    Message msg = CreatReproMessage();
    await client1.SendEventAsync(msg);
}

Message CreatReproMessage()
{
    Random rnd = new Random();
    int len = 1024 * 128;
    byte[] bytes = new byte[len];
    rnd.NextBytes(bytes);
    Message msg = new Message(bytes);
    int numProp = 82;
    for (int i = 0; i < numProp; i++)
    {
        msg.Properties.Add(Guid.NewGuid().ToString(), new string('*', 256));
    }
    return msg;
}

private void ConnectionStatusChanged(ConnectionStatus status, ConnectionStatusChangeReason reason)
{
    Debug.WriteLine($"Status: {status} - Reason: {reason}");
}
```
## Console log of the issue:

Status: Connected - Reason: Connection_Ok
Status: Disconnected - Reason: Communication_Error
Exception thrown: 'Microsoft.Azure.Devices.Client.Exceptions.IotHubException' in System.Private.CoreLib.dll
An exception of type 'Microsoft.Azure.Devices.Client.Exceptions.IotHubException' occurred in System.Private.CoreLib.dll but was not handled in user code
Work must be complete in the same order as it was started. Expected work id: '2', actual work id: '9522'

**Comment by jenoola** [comment]
Got repro on both UWP and console, both DeviceClient and ModuleClient. Everything works fine if numProp is decreased to 81. Investigating

**Comment by jenoola** [comment]
The issue is observed with MQTT only

**Comment by jenoola** [comment]
@wickste  So, what happens is when MqttIotHubAdapter converts the message into MQTT PublishPacket, it populates Message.Properties into PublishPacket.TopicName. The max length of PublishPacket.TopicName is 64KiB by MQTT spec (this is separate from message payload and you cannot overcome this restriction). Hence, when TopicName string is longer than 64KiB, user gets the exception. The only two options available would be 1) to set less properties for the message or 2) switch to AMQP because it has no restriction for this.

**Comment by timtay-microsoft** [comment]
Closing this PR for a fresh PR in #1197

**Comment by az-iot-builder-01** [comment]
@wickste, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)

**Comment by timtay-microsoft** [comment]
The fix for this issue has been released as of device client version 1.21.4, so I'm closing this issue

**Comment by az-iot-builder-01** [comment]
@wickste, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
